File: app/src/main/java/com/mandarin/bcu/androidutil/io/asynchs/Downloader.java
package com.mandarin.bcu.androidutil.io.asynchs;

import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.AsyncTask;
import android.view.View;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.mandarin.bcu.R;
import com.mandarin.bcu.androidutil.StaticStore;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.lang.ref.WeakReference;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class Downloader extends AsyncTask<Void, Integer, Void> {

    private int size;
    private Map<String, Long> sizes = new HashMap<>();
    private ArrayList<Boolean> remover = new ArrayList<>();

    private String[] lan = {"/en/", "/jp/", "/kr/", "/zh/"};
    private String difffile = "Difficulty.txt";
    private String source;
    private String downloading;
    private String extracting;
    private final String path;
    private final String mpath;
    private ArrayList<String> fileneed;
    private ArrayList<String> musics;

    private final WeakReference<Activity> weakActivity;

    private ArrayList<String> purifyneed = new ArrayList<>();

    private File output = null;

    public Downloader(String path, ArrayList<String> fileneed, ArrayList<String> music, String downloading, String extracting, Activity context) {
        this.path = path;
        this.fileneed = fileneed;
        this.downloading = downloading;
        this.extracting = extracting;
        this.weakActivity = new WeakReference<>(context);
        this.musics = music;
        source = path + "lang";
        mpath = path + "music/";
    }

    @Override
    protected void onPreExecute() {
        super.onPreExecute();
        Activity activity = weakActivity.get();

        if (activity == null) return;

        ProgressBar prog = activity.findViewById(R.id.downprog);
        TextView state = activity.findViewById(R.id.downstate);
        prog.setIndeterminate(true);
        state.setText(R.string.down_check_file);
    }

    @Override
    protected Void doInBackground(Void... voids) {
        String url;
        URL link;
        HttpURLConnection connection;
        String urls = "https://github.com/battlecatsultimate/bcu-resources/blob/master/resources/assets/";
        String RAW = "?raw=true";
        for (int i = 0; i < fileneed.size(); i++) {
            try {
                File f = new File(path, fileneed.get(i) + ".zip");

                if (!f.exists()) continue;

                url = urls + fileneed.get(i) + ".zip" + RAW;
                link = new URL(url);

                connection = (HttpURLConnection) link.openConnection();
                size = connection.getContentLength();

                sizes.put(fileneed.get(i), (long) size);

                connection.disconnect();
            } catch (MalformedURLException e) {
                e.printStackTrace();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }

        purify(sizes);

        int number;
        if(purifyneed.contains("Language") && purifyneed.contains("Music")) {
            number = purifyneed.size() - 2;
        } else if (purifyneed.contains("Language") || purifyneed.contains("Music")) {
            number = purifyneed.size() - 1;
        } else {
            number = purifyneed.size();
        }
        long total;
        for (int i = 0; i < number; i++) {
            try {
                url = urls + purifyneed.get(i) + ".zip" + RAW;
                link = new URL(url);
                connection = (HttpURLConnection) link.openConnection();
                connection.setRequestMethod("GET");
                connection.connect();

                size = connection.getContentLength();

                connection.getResponseCode();


                output = new File(path, purifyneed.get(i) + ".zip");
                File pathes = new File(path);

                if (!pathes.exists()) {
                    pathes.mkdirs();
                }

                if (!output.exists()) {
                    output.createNewFile();
                }

                FileOutputStream fos = new FileOutputStream(output);
                InputStream is = connection.getInputStream();

                byte[] buffer = new byte[1024];
                int len1;
                total = 0;

                while ((len1 = is.read(buffer)) != -1) {
                    total += Math.abs(len1);
                    int progress = (int) (total * 100 / size);
                    publishProgress(progress, i);
                    fos.write(buffer, 0, len1);
                }

                connection.disconnect();
            } catch (MalformedURLException e) {
                e.printStackTrace();
                output = null;
            } catch (IOException e) {
                e.printStackTrace();
                output = null;
            }

            if (output != null) {
                remover.add(false);
            }
        }

        System.out.println(musics);

        if(purifyneed.contains("Music")) {
            if(!musics.isEmpty()) {
                try {
                    for(int i = 0; i < musics.size(); i++) {
                        String murl = "https://github.com/battlecatsultimate/bcu-resources/raw/master/resources/music/";
                        String mfile = musics.get(i);

                        link = new URL(murl+mfile);

                        connection = (HttpURLConnection) link.openConnection();
                        connection.setRequestMethod("GET");
                        connection.connect();

                        connection.getResponseCode();

                        size = connection.getContentLength();

                        output = new File(mpath,mfile);

                        File mf = output.getParentFile();

                        if(!mf.exists()) {
                            mf.mkdirs();
                        }

                        if(!output.exists()) {
                            output.createNewFile();
                        }

                        FileOutputStream fos = new FileOutputStream(output);
                        InputStream is = connection.getInputStream();

                        byte [] buffer = new byte[1024];
                        int len1;
                        total = 0;

                        while((len1 = is.read(buffer)) != -1) {
                            total += len1;

                            int progress = 0;

                            if(size != 0)
                                progress = (int) (total*100/size);

                            publishProgress(progress,50,i);

                            fos.write(buffer,0,len1);
                        }

                        connection.disconnect();

                        fos.close();
                        is.close();
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }

            purifyneed.remove("Music");
            fileneed.remove("Music");
        }

        if (purifyneed.contains("Language")) {
            try {
                String lurl = "https://raw.githubusercontent.com/battlecatsultimate/bcu-resources/master/resources/lang";
                String durl = lurl + "/" + difffile;
                link = new URL(durl);
                connection = (HttpURLConnection) link.openConnection();
                connection.setRequestMethod("GET");
                connection.connect();

                connection.getResponseCode();

                output = new File(source + "/", difffile);

                File dpath = output.getParentFile();

                if (!dpath.exists()) {
                    dpath.mkdirs();
                }

                if (!output.exists()) {
                    output.createNewFile();
                }

                FileOutputStream dfos = new FileOutputStream(output);
                InputStream dis = connection.getInputStream();

                byte[] buffer = new byte[1024];
                int len1;
                total = 0;

                while ((len1 = dis.read(buffer)) != -1) {
                    total += len1;
                    int progress = 0;
                    if (size != 0)
                        progress = (int) (total * 100 / size);
                    publishProgress(progress, 100);
                    dfos.write(buffer, 0, len1);
                }

                connection.disconnect();
                dis.close();
                dfos.close();

                for (String s1 : lan) {
                    for (String s : StaticStore.langfile) {
                        String langurl = lurl + s1 + s + RAW;
                        link = new URL(langurl);
                        connection = (HttpURLConnection) link.openConnection();
                        connection.setRequestMethod("GET");
                        connection.connect();

                        connection.getResponseCode();

                        output = new File(source + s1, s);

                        File pathes = new File(source + s1);

                        if (!pathes.exists()) {
                            pathes.mkdirs();
                        }

                        if (!output.exists()) {
                            output.createNewFile();
                        }

                        FileOutputStream fos = new FileOutputStream(output);
                        InputStream is = connection.getInputStream();

                        buffer = new byte[1024];
                        total = 0;

                        while ((len1 = is.read(buffer)) > 0) {
                            total += len1;
                            int progress = 0;
                            if (size != 0)
                                progress = (int) (total * 100 / size);
                            publishProgress(progress, 100);
                            fos.write(buffer, 0, len1);
                        }

                        connection.disconnect();
                        fos.close();
                        is.close();
                    }

                    purifyneed.remove("Language");
                    fileneed.remove("Language");

                    StaticStore.unitlang = 1;
                    StaticStore.enemeylang = 1;
                    StaticStore.stagelang = 1;
                    StaticStore.maplang = 1;
                }
            } catch (IOException e) {
                output = null;
                e.printStackTrace();
            } finally {
                if (!(output == null)) {
                    fileneed.remove("Language");
                    purifyneed.remove("Language");
                }
            }
        }

        return null;
    }

    @Override
    protected void onProgressUpdate(Integer... values) {
        Activity activity = weakActivity.get();

        if (activity == null) return;

        ProgressBar prog = activity.findViewById(R.id.downprog);
        TextView state = activity.findViewById(R.id.downstate);

        if (prog.isIndeterminate()) {
            prog.setIndeterminate(false);
        }
        prog.setProgress(values[0]);
        if (values[1] != 100 && values[1] != 50) {
            state.setText(downloading + purifyneed.get(values[1]));
        } else if(values[1] == 50) {
            String mdownload = activity.getString(R.string.down_state_music)+musics.get(values[2]);

            state.setText(mdownload);
        } else {
            state.setText(downloading + "Language Files");
        }
    }

    @Override
    protected void onPostExecute(Void result) {
        Activity activity = weakActivity.get();

        if (activity == null) return;

        ProgressBar prog = activity.findViewById(R.id.downprog);
        TextView state = activity.findViewById(R.id.downstate);
        Button retry = activity.findViewById(R.id.retry);

        ArrayList<String> results = new ArrayList<>();

        if (remover.size() < purifyneed.size()) {
            for (int i = 0; i < purifyneed.size() - remover.size(); i++) {
                remover.add(true);
            }
        }

        for (int i = 0; i < remover.size(); i++) {
            if (remover.get(i)) {
                results.add(fileneed.get(i));
            }
        }

        purifyneed = results;

        if (prog.isIndeterminate()) {
            prog.setIndeterminate(false);
            prog.setProgress(100);
        }

        SharedPreferences preferences = activity.getSharedPreferences(StaticStore.CONFIG, Context.MODE_PRIVATE);

        if (!purifyneed.isEmpty() || output == null) {
            if (purifyneed.isEmpty()) {
                state.setText(R.string.down_state_ok);
                new Unzipper(path, fileneed, extracting, activity, preferences.getBoolean("upload", false) || preferences.getBoolean("ask_upload", true)).execute();
            } else {
                state.setText(R.string.down_state_no);
                retry.setVisibility(View.VISIBLE);
            }
        } else {
            state.setText(R.string.down_state_ok);
            new Unzipper(path, fileneed, extracting, activity, preferences.getBoolean("upload", false) || preferences.getBoolean("ask_upload", true)).execute();
        }
    }

    private void purify(Map<String, Long> size) {
        Activity activity = weakActivity.get();

        if (size == null || size.size() == 0) {
            purifyneed.addAll(fileneed);
            return;
        }

        if (activity == null) return;

        purifyneed = new ArrayList<>();
        ArrayList<Integer> result = new ArrayList<>();

        for (int i = 0; i < fileneed.size(); i++) {
            File f = new File(path, fileneed.get(i) + ".zip");

            if (f.exists()) {
                if (f.length() != size.get(fileneed.get(i))) {
                    result.add(i);
                }
            } else {
                result.add(i);
            }
        }

        if (!result.isEmpty()) {
            for (int i = 0; i < result.size(); i++) {
                purifyneed.add(fileneed.get(result.get(i)));
            }
        }
    }
}

File: app/src/main/java/com/mandarin/bcu/androidutil/io/asynchs/Downloader.kt
package com.mandarin.bcu.androidutil.io.asynchs

import android.app.Activity
import android.content.Intent
import android.os.AsyncTask
import android.view.View
import android.widget.Button
import android.widget.ProgressBar
import android.widget.TextView
import com.mandarin.bcu.CheckUpdateScreen
import com.mandarin.bcu.R
import com.mandarin.bcu.androidutil.StaticStore
import com.mandarin.bcu.androidutil.io.ErrorLogWriter
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.lang.ref.WeakReference
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL
import java.util.*
import kotlin.math.abs

class Downloader(private val fileneed: ArrayList<String>, music: ArrayList<String>, context: Activity) : AsyncTask<Void?, Int?, Void?>() {
    private var size = 0
    private val sizes: MutableMap<String, Long> = HashMap()
    private val remover = ArrayList<Boolean>()
    private val lan = arrayOf("/en/", "/jp/", "/kr/", "/zh/")
    private val difffile = "Difficulty.txt"
    private val musics: ArrayList<String> = music
    private val weakActivity: WeakReference<Activity> = WeakReference(context)
    private var purifyneed = ArrayList<String>()
    private var output: File? = null

    override fun onPreExecute() {
        super.onPreExecute()
        val activity = weakActivity.get() ?: return
        val prog = activity.findViewById<ProgressBar>(R.id.downprog)
        val state = activity.findViewById<TextView>(R.id.downstate)
        prog.isIndeterminate = true
        state.setText(R.string.down_check_file)
    }

    override fun doInBackground(vararg voids: Void?): Void? {
        val a = weakActivity.get() ?: return null

        var url: String
        var link: URL
        var connection: HttpURLConnection
        val urls = "https://github.com/battlecatsultimate/bcu-assets/blob/master/assets/"
        val raw = "?raw=true"

        val assetPath = StaticStore.getExternalAsset(a)+"assets/"

        for (i in fileneed.indices) {
            try {
                val f = File(assetPath, fileneed[i] + ".asset.bcuzip")
                if (!f.exists()) continue
                url = urls + fileneed[i] + ".asset.bcuzip" + raw
                link = URL(url)
                connection = link.openConnection() as HttpURLConnection
                size = connection.contentLength
                sizes[fileneed[i]] = size.toLong()
                connection.disconnect()
            } catch (e: MalformedURLException) {
                e.printStackTrace()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }

        purify(sizes)

        val number: Int = if (purifyneed.contains("Language") && purifyneed.contains("Music")) {
            purifyneed.size - 2
        } else if (purifyneed.contains("Language") || purifyneed.contains("Music")) {
            purifyneed.size - 1
        } else {
            purifyneed.size
        }

        var total: Long

        for (i in 0 until number) {
            try {
                url = urls + purifyneed[i] + ".asset.bcuzip" + raw

                link = URL(url)

                connection = link.openConnection() as HttpURLConnection
                connection.requestMethod = "GET"
                connection.connect()

                size = connection.contentLength

                publishProgress(-1,size)

                connection.responseCode

                output = File(assetPath, purifyneed[i] + ".asset.bcuzip")

                val pathes = File(assetPath)

                if (!pathes.exists()) {
                    pathes.mkdirs()
                }
                if (!output!!.exists()) {
                    output!!.createNewFile()
                }

                val fos = FileOutputStream(output)

                val `is` = connection.inputStream

                val buffer = ByteArray(1024)

                var len1: Int

                total = 0

                while (`is`.read(buffer).also { len1 = it } != -1) {
                    total += abs(len1).toLong()

                    val progress = total.toInt()

                    publishProgress(progress, i)
                    fos.write(buffer, 0, len1)
                }

                connection.disconnect()
            } catch (e: MalformedURLException) {
                ErrorLogWriter.writeLog(e, StaticStore.upload, a)

                output = null
            } catch (e: IOException) {
                ErrorLogWriter.writeLog(e, StaticStore.upload, a)

                output = null
            }
            if (output != null) {
                remover.add(false)
            }
        }

        println(musics)

        if (purifyneed.contains("Music")) {
            val musicPath = StaticStore.getExternalAsset(a) + "music/"

            if (musics.isNotEmpty()) {
                try {
                    for (i in musics.indices) {
                        val murl = "https://github.com/battlecatsultimate/bcu-resources/raw/master/resources/music/"
                        val mfile = musics[i]

                        link = URL(murl + mfile)

                        connection = link.openConnection() as HttpURLConnection
                        connection.requestMethod = "GET"
                        connection.connect()
                        connection.responseCode

                        size = connection.contentLength

                        publishProgress(-1,size)

                        output = File(musicPath, mfile)

                        val mf = output!!.parentFile ?: return null

                        if (!mf.exists()) {
                            mf.mkdirs()
                        }

                        if (!output!!.exists()) {
                            output!!.createNewFile()
                        }

                        val fos = FileOutputStream(output)
                        val `is` = connection.inputStream
                        val buffer = ByteArray(1024)
                        var len1: Int
                        total = 0
                        while (`is`.read(buffer).also { len1 = it } != -1) {
                            total += len1.toLong()
                            var progress = 0
                            if (size != 0) progress = total.toInt()
                            publishProgress(progress, 50, i)
                            fos.write(buffer, 0, len1)
                        }
                        connection.disconnect()
                        fos.close()
                        `is`.close()
                    }
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
            purifyneed.remove("Music")
            fileneed.remove("Music")
        }
        if (purifyneed.contains("Language")) {
            val langPath = StaticStore.getExternalAsset(a) + "lang/"

            try {
                val lurl = "https://raw.githubusercontent.com/battlecatsultimate/bcu-resources/master/resources/lang"
                val durl = "$lurl/$difffile"
                link = URL(durl)
                connection = link.openConnection() as HttpURLConnection
                connection.requestMethod = "GET"
                connection.connect()
                connection.responseCode
                output = File(langPath, difffile)
                val dpath = output!!.parentFile ?: return null
                if (!dpath.exists()) {
                    dpath.mkdirs()
                }
                if (!output!!.exists()) {
                    output!!.createNewFile()
                }
                val dfos = FileOutputStream(output)
                val dis = connection.inputStream
                var buffer = ByteArray(1024)
                var len1: Int
                total = 0
                while (dis.read(buffer).also { len1 = it } != -1) {
                    total += len1.toLong()
                    var progress = 0
                    if (size != 0) progress = (total * 100 / size).toInt()
                    publishProgress(progress, 100)
                    dfos.write(buffer, 0, len1)
                }
                connection.disconnect()
                dis.close()
                dfos.close()

                for (s1 in lan) {
                    for (s in StaticStore.langfile) {
                        val langurl = lurl + s1 + s + raw
                        link = URL(langurl)
                        connection = link.openConnection() as HttpURLConnection
                        connection.requestMethod = "GET"
                        connection.connect()
                        connection.responseCode
                        output = File(langPath + s1, s)
                        val pathes = File(langPath + s1)
                        if (!pathes.exists()) {
                            pathes.mkdirs()
                        }
                        if (!output!!.exists()) {
                            output!!.createNewFile()
                        }
                        val fos = FileOutputStream(output)
                        val `is` = connection.inputStream
                        buffer = ByteArray(1024)
                        total = 0
                        while (`is`.read(buffer).also { len1 = it } > 0) {
                            total += len1.toLong()
                            var progress = 0
                            if (size != 0) progress = (total * 100 / size).toInt()
                            publishProgress(progress, 100)
                            fos.write(buffer, 0, len1)
                        }
                        connection.disconnect()
                        fos.close()
                        `is`.close()
                    }
                    purifyneed.remove("Language")
                    fileneed.remove("Language")
                    StaticStore.unitlang = 1
                    StaticStore.enemeylang = 1
                    StaticStore.stagelang = 1
                    StaticStore.maplang = 1
                }
            } catch (e: IOException) {
                output = null
                e.printStackTrace()
            } finally {
                if (output != null) {
                    fileneed.remove("Language")
                    purifyneed.remove("Language")
                }
            }
        }
        return null
    }

   override fun onProgressUpdate(vararg values: Int?) {
        val activity = weakActivity.get() ?: return
        val prog = activity.findViewById<ProgressBar>(R.id.downprog)
        val state = activity.findViewById<TextView>(R.id.downstate)

        if (prog.isIndeterminate) {
            prog.isIndeterminate = false
        } else if(values[1] == 100) {
            prog.isIndeterminate = true
        }

        if(values[0] == -1) {
            prog.max = values[1] ?: 100
            return
        }

        prog.progress = values[0] ?: 0
        if (values[1] != 100 && values[1] != 50) {
            val t = activity.getString(R.string.down_state_doing) + purifyneed[values[1] ?: 0]
            state.text = t
        } else if (values[1] == 50) {
            val t = activity.getString(R.string.down_state_music) + musics[values[2] ?: 0]
            state.text = t
        } else {
            val t = activity.getString(R.string.down_state_doing) + "Language Files"
            state.text = t
        }
    }

    override fun onPostExecute(result: Void?) {
        val activity = weakActivity.get() ?: return
        val prog = activity.findViewById<ProgressBar>(R.id.downprog)
        val state = activity.findViewById<TextView>(R.id.downstate)
        val retry = activity.findViewById<Button>(R.id.retry)
        val results = ArrayList<String>()
        if (remover.size < purifyneed.size) {
            for (i in 0 until purifyneed.size - remover.size) {
                remover.add(true)
            }
        }
        for (i in remover.indices) {
            if (remover[i]) {
                results.add(fileneed[i])
            }
        }
        purifyneed = results

        if (prog.isIndeterminate) {
            prog.isIndeterminate = false
            prog.progress = 100
        }

        if (purifyneed.isNotEmpty() || output == null) {
            if (purifyneed.isEmpty()) {
                state.setText(R.string.down_state_ok)

                val intent = Intent(activity, CheckUpdateScreen::class.java)
                StaticStore.clear()
                activity.startActivity(intent)
                activity.finish()
            } else {
                state.setText(R.string.down_state_no)
                retry.visibility = View.VISIBLE
            }
        } else {
            state.setText(R.string.down_state_ok)

            val intent = Intent(activity, CheckUpdateScreen::class.java)
            StaticStore.clear()
            activity.startActivity(intent)
            activity.finish()
        }
    }

    private fun purify(size: Map<String, Long>?) {
        val activity = weakActivity.get() ?: return

        val assetPath = StaticStore.getExternalAsset(activity)

        if (size == null || size.isEmpty()) {
            purifyneed.addAll(fileneed)
            return
        }

        purifyneed = ArrayList()

        val result = ArrayList<Int>()

        for (i in fileneed.indices) {
            val f = File(assetPath, fileneed[i] + ".zip")
            if (f.exists()) {
                if (f.length() != size[fileneed[i]]) {
                    result.add(i)
                }
            } else {
                result.add(i)
            }
        }
        if (result.isNotEmpty()) {
            for (i in result.indices) {
                purifyneed.add(fileneed[result[i]])
            }
        }
    }

}